ext/drx/drx.h

```diff
@@ -459,6 +459,49 @@ void
 drx_buf_insert_buf_memcpy(void *drcontext, drx_buf_t *buf, instrlist_t *ilist,
                           instr_t *where, reg_id_t dst, reg_id_t src, ushort len);
 
+DR_EXPORT
+/**
+ * Expands AVX2 gather and AVX-512 gather and scatter instructions to a sequence of
+ * equivalent scalar load and stores, mask register bit tests, and mask register bit
+ * updates.
+ *
+ * WARNING: This function is not fully supported yet. Do not use.
+ *
+ * WARNING: The added multi-instruction loop contains several control-transfer
+ * instructions and is not straight-line code, which can complicate subsequent analysis
+ * routines.
+ *
+ * WARNING: This function makes a call to drreg_init() and drreg_exit(), extending the
+ * number of slots while setting do_not_sum_slots set to false.
+ *
+ * The client must use the \p drmgr Extension to order its instrumentation in order to
+ * use this function.  This function must be called from the application-to-application
+ * ("app2app") stage (see drmgr_register_bb_app2app_event()).
+ *
+ * This transformation is deterministic, so the caller can return
+ * DR_EMIT_DEFAULT from its event.
+ *
+ * The *dq, *qd, *qq, *dpd, *qps, and *qpd opcodes are not supported in 32-bit mode. In
+ * this case, the function will return true, but no expansion will occur.
```

Comment:
Why return true if it's failing to do what it's asked to do?

Reply:
I sticked to what drutil_expand_rep_string() does, e.g. for ARM. The rationale would be that the function was not called in error (e.g. from wrong phase), but an expansion didn't occur. If I change such that the return value determines whether an expansion occurred, I would get rid of the "OUT expanded". WDYT?

Reply:
Not following: drutil_expand_rep_string() for ARM *should* do nothing and return true.  I assume these opcodes listed here are scatter/gather opcodes and do need expansion and that expansion is simply not yet implemented, right?  So this function is failing for those.  Better to let the user know there's a problem, instead of the user gathering address traces and assuming they're correct and doing tons of work based on them and much later finding out they were wrong...

Reply:
>> Better to let the user know there's a problem
expanded is false in this case.

Up until now, the semantics of a _false_ return value was "error", like drreg error, wrong phase, called with scatter/gather instruction that doesn't make sense like wrong operands. Return _true_ would mean, "usage ok", like called with an instruction that isn't a scatter/gather instruction. I considered calling it with unsupported scatter/gather instructions like 32-bit qword ones to be "usage ok", but expanded will be false. You are saying you want to treat calling the expansion with a 32-bit qword scatter/gather like an error.  Fine with me! I changed the "XXX" into a "FIXME".

Done.

Reply:
The return value is whether the function was *successful* in executing what the user asked it to do.  If the function failed to expand a scatter/gather instruction, that's failure and it should return false.  The return value is not limited to user errors.

The user should not be given the burden of checking whether it passed a scatter/gather instruction, yet the optional "expanded" value was false, to figure out that there was a problem!  That is what the return value is for!